### PR TITLE
Use registers and instruction suffixes matching operand sizes

### PR DESCRIPTION
**Pull request by jspam**

### CODE REVIEW

File: src/main/scala/mjis/asm/Instruction.scala [hunk 1]
```diff
@@ -20,32 +21,41 @@ object AMD64Registers {
...
+object Instruction {
+  def suffixForSize(sizeBytes: Int): String = sizeBytes match {
+    case 8 => "q"
+    case 4 => "l"
+    case 2 => "w"
+    case 1 => "b"
+  }
+}
+
+abstract class Operand(val sizeBytes: Int)
```
Comment (Kha) [inline src/main/scala/mjis/asm/Instruction.scala]: No `sealed` anymore?

File: src/main/scala/mjis/CodeGenerator.scala [hunk 2]
```diff
@@ -36,9 +35,18 @@ class CodeGenerator(a: Unit) extends Phase[AsmProgram] {
     resultProgram
   }
 
-  def intConstOp(i: Int): Operand = new ConstOperand(i, Mode.getIs.getSizeBytes)
-  def regOp(node: Node): RegisterOperand = new RegisterOperand(node.idx, node.getMode.getSizeBytes)
-  implicit def regOp(regNr: Int): RegisterOperand = new RegisterOperand(regNr, Registers(regNr).sizeBytes)
+  def intConstOp(i: Int): Operand = ConstOperand(i, Mode.getIs.getSizeBytes)
+  def regOp(node: Node): RegisterOperand = RegisterOperand(node.idx, node match {
+    case n : firm.nodes.Call =>
```
Comment (Kha) [inline src/main/scala/mjis/CodeGenerator.scala]: nit: Shorten to `nodes.`?

Comment (jspam) [inline src/main/scala/mjis/CodeGenerator.scala]: Done

File: src/main/scala/mjis/RegisterAllocator.scala [hunk 3]
```diff
@@ -61,14 +63,14 @@ class FunctionRegisterAllocator(function: AsmFunction) {
...
-                instrList.insert(i, Movq(c, actualRegister).withComment(s" - Load constant ${instr.comment}"))
+                val actualRegister = getRegister(c.sizeBytes)
+                instrList.insert(i, Mov(c, actualRegister).withComment(s" - Load constant ${instr.comment}"))
                 i += 1
                 instr.operands(idx) = actualRegister
-              case r@RegisterOffsetOperand(oldRegNr, _, _) =>
-                val actualRegister = getRegister()
-                reload(oldRegNr, actualRegister)
-                instr.operands(idx) = r.copy(regNr = actualRegister)
+              case r@RegisterOffsetOperand(reg@RegisterOperand(_, sizeBytes), _, _) =>
```
Comment (Kha) [inline src/main/scala/mjis/RegisterAllocator.scala]: I think I'd prefer using `reg.sizeBytes` :) .

File: src/main/scala/mjis/asm/Instruction.scala [hunk 4]
```diff
@@ -71,21 +81,53 @@ abstract class Instruction(private val operandsWithSpec: (Operand, OperandSpec)*
   var stackPointerDisplacement: Int = 0
   val operands = ListBuffer[Operand](operandsWithSpec.map(_._1):_*)
   val operandSpecs = Seq[OperandSpec](operandsWithSpec.map(_._2):_*)
+  def suffix = ""
```
Comment (Kha) [inline src/main/scala/mjis/asm/Instruction.scala]: Default implementation: Assert that all operands have the same byteSize and use that one?

**Comment by Kha** [comment]
Modulo comments, LGTM

**Pull request merged**